### PR TITLE
Changes head IK to use a template function

### PR DESCRIPTION
**Pull request by ysims**
So pretty sure both `calculateCameraLookJoints` and `calculateHeadJoints` are the same functions but one is float and the other is double, although they have some strange stylistic differences as well.
This just removes `calculateCameraLookJoints` and changes so we just use `calculateHeadJoints` in place of that.

Please tell me if this is good or if I need to sleep 😬

### CODE REVIEW

File: shared/tests/utility/motion/TestKinematics.cpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ TEST_CASE("Test the Head kinematics", "[utility][motion][kinematics][head]") {
         INFO("Testing with the random vector, " << camera_vector.transpose());
 
         std::vector<std::pair<ServoID, double>> angles =
```
Comment (KipHamiltons) [inline shared/tests/utility/motion/TestKinematics.cpp:59]: Should you change this to use `float`s too?

File: module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp [hunk 2]
```diff
@@ -479,7 +479,7 @@ namespace module::behaviour::skills {
         // Rotate target angles to World space
         Eigen::Vector3d lookVector = headToIMUSpace * lookVectorFromHead;
         // Compute inverse kinematics for head direction angles
-        std::vector<std::pair<ServoID, double>> goalAngles = calculateCameraLookJoints(lookVector);
+        std::vector<std::pair<ServoID, float>> goalAngles = calculateHeadJoints(lookVector);
```
Comment (abihall) [inline module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp:482]: Why was this changed to a float?

Comment (ysims) [reply module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp:482]: Because `calculateHeadJoints` returns a float

File: shared/utility/motion/InverseKinematics.cpp [hunk 3]
```diff
@@ -337,23 +337,14 @@ namespace utility::motion::kinematics {
         joints.insert(joints.end(), joints2.begin(), joints2.end());
         return joints;
     }
-    std::vector<std::pair<ServoID, double>> calculateCameraLookJoints(const Eigen::Vector3d& cameraUnitVector) {
-        std::vector<std::pair<ServoID, double>> positions;
-        positions.push_back(std::make_pair(ServoID::HEAD_YAW, std::atan2(cameraUnitVector.y(), cameraUnitVector.x())));
+    std::vector<std::pair<ServoID, float>> calculateHeadJoints(const Eigen::Vector3f& cameraUnitVector) {
+        std::vector<std::pair<ServoID, float>> positions;
```
Comment (Bidski) [inline shared/utility/motion/InverseKinematics.cpp:341]: You could always template this function to allow for some flexibility
```suggestion
    template <typename Scalar>
    std::vector<std::pair<ServoID, Scalar>> calculateHeadJoints(const Eigen::Matrix<Scalar, 3, 1>& cameraUnitVector) {
        std::vector<std::pair<ServoID, Scalar>> positions;
```

**Review by Bidski — CHANGES_REQUESTED** [review]

File: module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp [hunk 4]
```diff
@@ -479,7 +479,7 @@ namespace module::behaviour::skills {
         // Rotate target angles to World space
         Eigen::Vector3d lookVector = headToIMUSpace * lookVectorFromHead;
         // Compute inverse kinematics for head direction angles
-        std::vector<std::pair<ServoID, double>> goalAngles = calculateCameraLookJoints(lookVector);
+        std::vector<std::pair<ServoID, double>> goalAngles = calculateHeadJoints(lookVector);
 
         Eigen::Vector2d result;
         for (auto& angle : goalAngles) {
```
Comment (TheCoolKidsTable) [inline module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp:485]: While we are here, is this really the best way to extract the angles from std::vector<std::pair<ServoID, double>> into a Vector2d. Seems like a very strange way to go about it?

Comment (ysims) [reply module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp:485]: I can't think of any better way, since we don't really know where the head servos are in the vector. Did you have an idea?

File: shared/utility/motion/InverseKinematics.hpp [hunk 5]
```diff
@@ -68,11 +68,16 @@ namespace utility::motion::kinematics {
         const Eigen::Affine3f& leftTarget,
         const Eigen::Affine3f& rightTarget);
 
-    [[nodiscard]] std::vector<std::pair<ServoID, double>> calculateCameraLookJoints(
-        const Eigen::Vector3d& cameraUnitVector);
-
-    [[nodiscard]] std::vector<std::pair<ServoID, float>> calculateHeadJoints(Eigen::Vector3f cameraUnitVector);
-
+    template <typename Scalar>
+    std::vector<std::pair<ServoID, Scalar>> calculateHeadJoints(const Eigen::Matrix<Scalar, 3, 1>& cameraUnitVector) {
```
[the remark this answers is not in the document]
Comment (ysims) [reply shared/utility/motion/InverseKinematics.hpp:72]: There's already a test for all the kinematics stuff?

[the remark this answers is not in the document]
Comment (KipHamiltons) [reply shared/utility/motion/InverseKinematics.hpp:72]: Ooft. My bad 😅

File: shared/utility/motion/InverseKinematics.hpp [hunk 6]
```diff
@@ -68,11 +68,16 @@ namespace utility::motion::kinematics {
...
+        std::vector<std::pair<ServoID, Scalar>> positions;
+        positions.push_back(std::make_pair(ServoID::HEAD_YAW, atan2(cameraUnitVector.y(), cameraUnitVector.x())));
+        positions.push_back(std::make_pair(ServoID::HEAD_PITCH,
+                                           std::atan2(-cameraUnitVector.z(),
+                                                      std::sqrt(cameraUnitVector.x() * cameraUnitVector.x()
+                                                                + cameraUnitVector.y() * cameraUnitVector.y()))));
+        return positions;
+    }
```
Comment (KipHamiltons) [inline shared/utility/motion/InverseKinematics.hpp:80]: This probably breaks formatting and I might have mismatched brackets :p
```suggestion
    [[nodiscard]] std::vector<std::pair<ServoID, Scalar>> calculateHeadJoints(
        const Eigen::Matrix<Scalar, 3, 1>& cameraUnitVector) {
        std::vector<std::pair<ServoID, Scalar>> positions{};
        positions.emplace_back(ServoID::HEAD_YAW, atan2(cameraUnitVector.y(), cameraUnitVector.x()));
        positions.emplace_back(ServoID::HEAD_PITCH,
                               std::atan2(-cameraUnitVector.z(),
                                          std::sqrt(cameraUnitVector.x() * cameraUnitVector.x()
                                                    + cameraUnitVector.y() * cameraUnitVector.y())));
        return positions;
    }
```

**Review by KipHamiltons — CHANGES_REQUESTED** [review]
`[[nodiscard]]` :triumph: also `emplace_back` pls (you could use an aggregate initialisation of the vector if you wanted too - as in, braces list of elements).

File: module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp (hunk 4 above)
Comment (KipHamiltons) [reply module/behaviour/skills/HeadBehaviourSoccer/src/HeadBehaviourSoccer.cpp:485]: Make an issue if it's worth making an issue

**Review by KipHamiltons — APPROVED** [review]
LGTM

**Review by Bidski — APPROVED** [review]
Dont see anything offensive
